mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -521,11 +528,74 @@ static bool isUnfusedKernelStore(ThreadwiseWriteAllOp store) {
   return ret;
 }
 
+LogicalResult
+ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+                                      PatternRewriter &rewriter) const {
+  Location loc = reduceOp.getLoc();
+  SmallVector<TransformMapAttr, 4> views;
+  ThreadwiseWriteAllOp threadwiseWriteOp =
+      traceToThreadwiseWrite(reduceOp.getIn(), views);
+  if (!threadwiseWriteOp) {
+    return rewriter.notifyMatchFailure(
+        reduceOp, "rock.reduce input is not leading to ThreadwiseWriteAllOp of "
+                  "the previous op.");
+  }
+  if (reduceOp.getReduceMethod() != ReduceMethod::Sum) {
+    // We are failing the pass here because rock.reduce appearing here means
+    // we are committed to fusion and this is case, we cant handle (so far) in
+    // this or a later pass.
+    return reduceOp.emitError("We only support sum reductions.!");
+  }
+  if (threadwiseWriteOp.getStoreMethod() != rock::StoreMethod::Set) {
+    // We are failing the pass here because another rock.reduce appearing here
+    // means we are committed to fusion and this is case, we cant handle (so
+    // far) in this or a later pass.
+    return reduceOp.emitError("Another reduction op is not able to be fused "
+                              "with a prior reduction op.");
+  }
+
+  bool isUniqueReader;
+  LogicalResult checkResult = checkUniqueReader(
+      reduceOp.getIn().getDefiningOp(), reduceOp, isUniqueReader);
+  if (checkResult.failed()) {
+    return checkResult;
+  }
+  if (!isUniqueReader) {
+    threadwiseWriteOp = static_cast<ThreadwiseWriteAllOp>(
+        rewriter.clone(*threadwiseWriteOp.getOperation()));
+  }
+  int64_t reductionAxis = reduceOp.getAxisAttr().getInt();
+  TypedValue<ShapedType> redIn = reduceOp.getIn();
+  ArrayRef<int64_t> reduceInShape = redIn.getType().getShape();
+  TopDownTMBuilder dropReductionDim(rewriter, reduceInShape, loc);
+  for (uint32_t i = 0; i < reduceInShape.size(); ++i) {
+    if (i == reductionAxis) {
+      dropReductionDim.constDim("reduction_dim", i, 0, 1);
```

Comment:
Is `constDim` the right operation here? I'd think it might be `broadcast`?

The point of this map, as far as I understand it, is to map `(..., dR, ...) -> (... 0, ...)`, aka, it's the map where `dR` was broadcast.

Reply:
It is a fine line...

Id agree with broadcast if it was a load.

However, in a reduction store, we just force dR to be zero -- so constDim made slightly bit more sense..

Reply:
A broadcast does force the dimension to 0

I think a broadcast is more semantically correct and might help the vectorizer out.

Reply:
Ok, I ll buy the vectorizer argument if thats the case.

Reply:
Basically, the contiguous-dimensions checker understands broadcasts, but doesn't understand not using an input and then defining a 0 - it won't know the relationship between the argument and the result.

It'd take a somewhat contrived scenario for this to actually come up - mainly a reduction of length 1 - but still.

Reply:
ok done!